dss_client/src/dss_client.cpp

```diff
@@ -1323,7 +1340,90 @@ namespace dss {
 			printf("Exception caught in DeleteObject for %s\n", key_str.c_str());
 		}
 		return ret;
+	}
+
+	extern "C" int ListObjects(DSSClient c, char* prefix, char* delimit, char* keys, int cur_pg) 
+	{
+		Client *client = (Client*) c;
+		if (client == nullptr){
+			printf("the DSS client cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (keys == nullptr){
+			printf("the buffer to store keys cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (cur_pg == -1){ // for the first page of list
+			client->list_objs = client->GetObjects(prefix, delimit);
+		}
+		if (client->list_objs->GetObjKeys() < 0) { // reached the end of the pages for LIST
+			return END_OF_LIST;
+		}
+		std::string keys_concat = "";
+		for (std::string key: client->list_objs->GetPage()){
+			keys_concat += key + "\n"; // seperate each key by line assuming there is no "\n" in the object name or directory name or delimiter
+		}
+		if (keys_concat == "") return END_OF_LIST;
+		keys[keys_concat.length()] = '\0';
+		strncpy(keys, keys_concat.c_str(), keys_concat.length());
+		if (keys[keys_concat.length()] != '\0') {
+			printf("the buffer overflow for storing keys\n");
+			return FAILURE;
+		}
+		return cur_pg + 1;
+	}
 
+	extern "C" int DeleteAll(DSSClient c, char* prefix, char* delimit)
+	{
+		Client *client = (Client*) c;
+		if (client == nullptr){
+			printf("the DSS client cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (prefix == nullptr){
+			printf("the prefix cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		if (delimit == nullptr){
+			printf("the delimiter cannot be a null pointer.\n");
+			return FAILURE;
+		}
+		uint32_t pg_size = DSS_PAGINATION_DEFAULT;
+		int max_key_len = 1024;
+		char* keys = (char*) malloc(sizeof(char) * max_key_len * pg_size);
+		int pg;
+		while (true){
+			pg = ListObjects(client, prefix, delimit, keys, -1); // list the very first page
+			if (pg == FAILURE){
+				printf("ListObjects failed for prefix=\"%s\" and delimit=\"%s\".\n", prefix, delimit);
+				printf("DeleteAll failed for prefix=\"%s\" and delimit=\"%s\".\n", prefix, delimit);
+				free(keys);
+				return FAILURE;
+			}
+			if (pg == END_OF_LIST) break;
+			std::stringstream all_keys(keys);
```

Comment:
You can check whether strtok or stringstream is efficient and use it accordingly.